Resources/Storage/S3Storage.py

```diff
@@ -0,0 +1,748 @@
+
+# docker run --rm -it --privileged --name dirac-testing-host \
+#   -v /home/chaen/dirac/diracostar:/diracostar \
+#   -e DIRACOS_TARBALL_PATH=/diracostar \
+#   -e CI_PROJECT_DIR=/repo/DIRAC -e CI_REGISTRY_IMAGE=diracgrid \
+#   -v /var/run/docker.sock:/var/run/docker.sock -v $PWD:/repo/DIRAC -w /repo \
+#   diracgrid/docker-compose-dirac:latest bash \
+#   DIRAC/tests/CI/run_docker_setup.sh
+
+"""
+Configuration of an S3 storage
+Like others, but in protocol S3 add:
+
+SecureConnection: true if https, false otherwise
+Aws_access_key_id
+Aws_secret_access_key
+
+if the Aws variables are not defined, it will try to go throught the S3GW
+
+"""
+__RCSID__ = "$Id$"
+
+
+import copy
+import functools
+
+import os
+import requests
+
+
+import boto3
+from botocore.exceptions import ClientError
+
+from DIRAC import S_OK, S_ERROR, gLogger
+from DIRAC.Core.Utilities.Adler import fileAdler
+from DIRAC.Core.Utilities.Pfn import pfnparse
+from DIRAC.DataManagementSystem.Client.S3GWClient import S3GWClient
+from DIRAC.Resources.Storage.StorageBase import StorageBase
+
+
+LOG = gLogger.getSubLogger(__name__)
+
+
+def _extractKeyFromS3Path(meth):
+  """ Decorator to split an s3 "external" url (s3://server:port/bucket/path)
+      and return only the path part.
+  """
+
+  @functools.wraps(meth)
+  def extractKey(self, urls, *args, **kwargs):
+
+    # If set to False, we are already working with keys, so
+    # skip all the splitting
+    extractKeys = kwargs.pop('extractKeys', True)
+
+    keysToUrls = {}
+    keyArgs = {}
+
+    successful = {}
+    failed = {}
+
+    if extractKeys:
+      for url in urls:
+        res = pfnparse(url, srmSpecific=False)
+        if not res['OK']:
+          failed[url] = res['Message']
+          continue
+
+        splitURL = res['Value']
+        path = splitURL['Path'].lstrip('/')
+        key = os.path.join(path, splitURL['FileName'])
+        keysToUrls[key] = url
+        keyArgs[key] = urls[url]
+
+    else:
+      keyArgs = copy.copy(urls)
+
+    result = meth(self, keyArgs, *args, **kwargs)
+
+    if not result['OK']:
+      return result
+    # Restore original paths
+
+    for key in result['Value']['Failed']:
+      failed[keysToUrls.get(key, key)] = result['Value']['Failed'][key]
+    for key in result['Value']['Successful']:
+      successful[keysToUrls.get(key, key)] = result['Value']['Successful'][key]
+
+    result['Value'].update({"Successful": successful, "Failed": failed})
+    return result
+
+  return extractKey
+
+
+class S3Storage(StorageBase):
+  """
+  .. class:: StorageBase
+
+  """
+
+  pluginName = 'S3'
+
+  def __init__(self, storageName, parameters):
+
+    super(S3Storage, self).__init__(storageName, parameters)
+
+    self.isok = True
+
+    aws_access_key_id = parameters.get('Aws_access_key_id')
+    aws_secret_access_key = parameters.get('Aws_secret_access_key')
+    secureConnection = (parameters.get('SecureConnection', False) == 'True')
```

Comment:
ceph default is apparently `http` ...

Reply:
I don't think that should dictate what DIRAC's default it. Most setups have HTTPS enabled and people can edit the CS if theirs doesn't.

Reply:
the argument that people can edit their CS works in both direction :-D Anyway, I really don't care. To be honest, it's really because the default of CEPH and the mock I used is HTTP :-D But I can also change the tests to overwrite that

Reply:
> the argument that people can edit their CS works in both direction

Yes but my way is secure by default.

Reply:
I'm fine with that yes